index.js

```diff
@@ -172,80 +172,80 @@
+      x = has(path[i], x) && x[path[i]];
+      if (!x) return null;
+    }
+    return x;
   }
 
   //  functionName :: Function -> String
-  var functionName = 'name' in function f() {} ?
+  var functionName = 'name' in identity ?
     function functionName(f) { return f.name; } :
     /* istanbul ignore next */
     function functionName(f) {
       var match = /function (\w*)/.exec(f);
       return match == null ? '' : match[1];
     };
 
-  //  $ :: (String, Array TypeClass, StrMap (Array Location)) -> TypeClass
-  function $(_name, dependencies, requirements) {
-    function getBoundMethod(_name) {
-      var name = 'fantasy-land/' + _name;
-      return requirements[_name] === Constructor ?
-        function(typeRep) {
-          var f = funcPath([name], typeRep);
-          return f == null && typeof typeRep === 'function' ?
-            implPath([functionName(typeRep), name]) :
-            f;
-        } :
-        function(x) {
-          var isPrototype = x != null &&
-                            x.constructor != null &&
-                            x.constructor.prototype === x;
-          var m = null;
-          if (!isPrototype) m = funcPath([name], x);
-          if (m == null)    m = implPath([type(x), 'prototype', name]);
-          return m && m.bind(x);
-        };
-    }
+  function getBoundFunction(_name) {
+    var name = Prefix + _name;
+    return function(typeRep) {
+      var f = funcPath([name], typeRep);
+      return f == null && typeof typeRep === 'function' ?
+        implPath([functionName(typeRep), name]) :
+        f;
+    };
+  }
 
+  function getBoundMethod(_name) {
+    var name = Prefix + _name;
+    return function(x) {
+      var isPrototype = x != null &&
+                        x.constructor != null &&
+                        x.constructor.prototype === x;
+      var m = null;
+      if (!isPrototype) m = funcPath([name], x);
+      if (m == null)    m = implPath([type(x), 'prototype', name]);
+      return m && m.bind(x);
+    };
+  }
+
+  //  $ :: (String, Array TypeClass, Array String, Array String) -> TypeClass
+  function $(_name, dependencies, functionKeys, methodKeys) {
     var name = 'sanctuary-type-classes/' + _name;
-    var keys = Object.keys(requirements);
 
     var typeClass = TypeClass(name, dependencies, function(x) {
-      return keys.every(function(_name) {
-        var arg = requirements[_name] === Constructor ? x.constructor : x;
-        return getBoundMethod(_name)(arg) != null;
+      return functionKeys.every(function(k) {
+        return typeClass.functions[k](x.constructor) != null;
+      }) && methodKeys.every(function(k) {
+        return typeClass.methods[k](x) != null;
       });
     });
 
-    typeClass.methods = keys.reduce(function(methods, _name) {
-      methods[_name] = getBoundMethod(_name);
-      return methods;
-    }, {});
+    functionKeys.forEach(function(k) {
+      this[k] = getBoundFunction(k);
+    }, typeClass.functions = {});
```

Comment:
I could move `typeClass.functions = {}` to above, and replace `this` by `typeClass.functions` for readability.